python: Update testsuite to pass with distutils-C++ patch

bernhardkaindl opened the pull request:
Fix passing the python testsuite with spack install --test:

The distutils-C++ patch fixes missing support for CXXFLAGS and CPPFLAGS
for C++ compiles. This patch updates the distutils testsuite to expect
these additional flags to be passed to the C++ compiler and also covers
the other additions of the patch.

adamjstewart commented on var/spack/repos/builtin/packages/python/package.py:
```diff
@@ -195,6 +195,7 @@ class Python(AutotoolsPackage):
     patch('python-3.6.8-distutils-C++.patch', when='@3.6.8,3.7.2')
     patch('python-3.7.3-distutils-C++.patch', when='@3.7.3')
     patch('python-3.7.4+-distutils-C++.patch', when='@3.7.4:')
```
Shouldn't this also be done for 2.7.8 to 3.7.3?

bernhardkaindl replied:
Strictly said: No, because not all of these releases (2.7.8 to 3.7.3) have a distutils-C++ patch.

In practicality, it is not really worth to write testsuite patches for each of these individual versions, since none of them pass the rest of the python testsuite anyways in spack. The other testsuite fails already start coming when going just back to 3.7.9.

bernhardkaindl replied:
Actually, python@2.7.17:2.7.18 also have no other testsuite fails for me.

I've updated the testsuite for `python-2.7.17+-distutils-C++.patch` and found a bug in it:

This how the test suite would have to change when the bug is not fixed:
```py
         self.assertEqual(comp.exes['compiler'],
-                         'env_cc --sc-cflags --env-cflags --env-cppflags')
+                         'env_cc  --env-cflags --env-cppflags')
         self.assertEqual(comp.exes['compiler_so'],
-                         ('env_cc --sc-cflags '
+                         ('env_cc  '
                           '--env-cflags ''--env-cppflags --sc-ccshared'))
         self.assertEqual(comp.exes['compiler'],
-                         'sc_cc --sc-cflags')
+                         'sc_cc ')
         self.assertEqual(comp.exes['compiler_so'],
-                         'sc_cc --sc-cflags --sc-ccshared')
+                         'sc_cc  --sc-ccshared')
         self.assertEqual(comp.exes['compiler_cxx'],
-                         'sc_cxx')
+                         'sc_cxx ')
```
As shown, it removes the default, builtin cflags when calling the pure C compiler CC, which is wrong,
introducing a regression. The cause is that the patch removes reading the python distutils cflags value and sets both empty.

```py
-        (cc, cxx, cflags, ccshared, ldshared, so_ext, ar, ar_flags) = \
-            get_config_vars('CC', 'CXX', 'CFLAGS',
-                            'CCSHARED', 'LDSHARED', 'SO', 'AR',
-                            'ARFLAGS')
+        (cc, cxx, ccshared, ldshared, ldcxxshared, so_ext, ar, ar_flags) = \
+            get_config_vars('CC', 'CXX', 'CCSHARED', 'LDSHARED', 'LDCXXSHARED',
+                            'SO', 'AR', 'ARFLAGS')
+        cflags = ''
+        cxxflags = ''
``` 
The fix is obvious when comparing it with python-3.7.4+-distutils-C++.patch
```py
-        (cc, cxx, cflags, ccshared, ldshared, shlib_suffix, ar, ar_flags) = \
-            get_config_vars('CC', 'CXX', 'CFLAGS',
-                            'CCSHARED', 'LDSHARED', 'SHLIB_SUFFIX', 'AR', 'ARFLAGS')
+        (cc, cxx, cflags, ccshared, ldshared, ldcxxshared, shlib_suffix, ar, ar_flags) = \
+            get_config_vars('CC', 'CXX', 'CFLAGS', 'CCSHARED', 'LDSHARED', 'LDCXXSHARED',
+                            'SHLIB_SUFFIX', 'AR', 'ARFLAGS')
+
+        cxxflags = cflags
```
When handling CFLAGS the same as in python-3.7.4+-distutils-C++.patch, the issue is fixed.

I'll add the resulting fixup of the patch for 2.7.18 with the fixes testsuite patch for 2.7.18.

Pushed, now `spack install --test` passes for 3.8.11 and 2.7.18. @xjrc you might want to look at this comment specifically and the new fixup patch for 2.7.18.

xjrc replied:
@bernhardkaindl: Doing my best to recall my headspace when working on #17140, I recall throwing some of these patches together without the greatest deal of precision. Based on the PR comments, it looks like I took these patches more-or-less verbatim from [a patch set in the Python bug tracker](https://bugs.python.org/issue1222585). The patch set was never merged, so it doesn't surprise me that some of its constituents are imperfect. It looks like you were able to diagnose these imperfections and come up with a solution though, so thanks for looking into this and kudos!

bernhardkaindl replied:
Thanks, I appreciate that!

I wanted to see if I could get `spack install --test all` working for the core packages, so I found many things in the last days.

I'm not surprised of such 'gotchas' anymore after that, and as these patches didn't even update the `distutils` testsuite to pass, it's no surprise that some issues were missed when those patches were developed.

bernhardkaindl replied:
I did a little bit of history reading:

In 2015, Erik Hvatum (using gentoo) found the same issue already.
In 2017, Josef Kemetmüller applied Erik's fix to his patch for python3.x, but he didn't update the python2.7 patch.
That means, since 2017, the patch for "3.x" from Josef and its derivatives have the `CFLAGS` fix:
https://bugs.python.org/issue1222585#msg242706

Erik linked the gentoo bug report he opened for the dropped `CFLAGS` and there you see that these patches originate from Gentoo. Gentoo Developer Mike Gilbert didn't know the reason why they wiped out the cflags, it was before his time: https://bugs.gentoo.org/548776#c3

He was pretty sure it was intentional, but the reason might have been anything: Maybe they didn't want the `CFLAGS` of python itself used for module builds as well, but that is actually the intention of the python devs in doing so.

It should be safe to do what upstream Python decided to do, and we used the variant of keeping the `CFLAGS` in the 3.7.4+ patch series already.

@xjrc Thanks for approving the fix.

adamjstewart commented:
@xjrc added some of these patches and may also want to review.

xjrc commented on var/spack/repos/builtin/packages/python/python-2.7.17+-distutils-C++-fixup.patch:
```diff
@@ -0,0 +1,120 @@
+This patch fixes a regression intoduced by python-2.7.17+-distutils-C++.patch
+and updates the distutils testsuite to pass with 2.7.16:2.7.18:
```
I've never seen a comment like this in a patch file before; thanks for documenting this fix so thoroughly!

bernhardkaindl replied:
It was just a cut and paste from the comment above - where it looks a bit better and was nicer to edit.

But yes, I understand that it is good to have the documentation also right there in the patch.
Applying past experience of mine, and as I already wrote it for the review itself.

Adding it to the patch for the benefit of others (or one self later) without having to wade thru the pull request makes it easy to see what a few lines of patch do.

bernhardkaindl commented:
@spackbot run pipeline

bernhardkaindl commented:
@spackbot run pipeline

bernhardkaindl commented:
@spackbot run pipeline

bernhardkaindl committed:
python: Fix regression in python-2.7.17+-distutils-C++.patch

This commit also updates the python testsuite for the changes which
these python patches applied by this package make:
- python-2.7.17+-distutils-C++.patch
- python-3.7.4+-distutils-C++-testsuite.patch

This fixes spack install --test for these respecitve versions. Details:

The distutils-C++ patches fixes missing support for CXXFLAGS and CPPFLAGS
for C++ compiles. This commit updates the distutils testsuite to expect
these additional flags to be passed to the C++ compiler and also covers
the other additions of the python-3.7.4+-distutils-C++-testsuite.patch,
and (to a lesser extent) python-2.7.17+-distutils-C++.patch.

bernhardkaindl commented:
@spackbot run pipeline

adamjstewart commented:
Not sure why spackbot is broken, we're looking into this.

bernhardkaindl commented:
> Not sure why spackbot is broken, we're looking into this.

Last week, there was an CI issue in one of my PRs which wasn't yet updated to be based on the new develop commit of PR #22845 which changed the database to v6 format, where the old v5 spack of my branch wasn't able to deal with the v6 database from the CI setup it seems.

I've looked at what would be needed to be able to run v5 db spack with a $SPACK_ROOT directory which already has a v6 database. I submitted a quick take on that as PR #25890.

With that PR, the v5 and v6 spacks don't "spack find" (reindex)each other's `spec.yaml`/`spec.json` files yet, but both v5 and v6-db-based spack versions are able to run and use their own installed specs, misc_cache and database/index.json file.

I rebased and squashed my commits now on top of today's latest develop branch.

In case the test-bot looks at if the - "to be built" - commit is based on the code from PR #22845, this might unlock the CI then. At the least, it would have likely failed without the rebase to the latest develop.

bernhardkaindl commented:
@spackbot run pipeline

bernhardkaindl commented:
@spackbot run pipeline

bernhardkaindl commented:
@spackbot help

bernhardkaindl commented:
> I had a problem triggering the pipeline.

@scottwittenburg spackbot still has a problem triggering this pipeline.

Is your fix https://github.com/spack/spackbot/pull/54 (URL encode branch names in case of reserved chars) already deployed to github?

scottwittenburg commented:
@spackbot run pipeline

scottwittenburg commented:
> Is your fix [spack/spackbot#54](https://github.com/spack/spackbot/pull/54) (URL encode branch names in case of reserved chars) already deployed to github?

Looks like it is not.  Pinging @tgamblin as he may have to prioritize fixing the issue with spackbot `main` or task someone else with it.

scottwittenburg commented:
@zackgalbreath At least the spack-infrastructure fix is deployed and this confirms it's working, as last time I manually ran a pipeline on this PR, `gh-gl-sync` never noticed and it stayed backlogged.

zackgalbreath commented:
> @zackgalbreath At least the spack-infrastructure fix is deployed and this confirms it's working, as last time I manually ran a pipeline on this PR, `gh-gl-sync` never noticed and it stayed backlogged.

Glad to hear it, thanks for letting me know.

bernhardkaindl commented:
@skosukhin: spackbot also requested a review from you, I summarize:

@xjrc, who integrated the distutils patches reviewed my changes and approved them.

#### Summary:

##### For the distutils-C++ patch for python 3.7.+:

This PR only updates of the `distutils` *unit-tests*, so they pass when spack runs `spack install --test=root`.

##### For the distutils-C++ patch for python 2.7.16:2.7.18:

For these versions, this PR adds a small change, summarized in the new patch file like this:
```
This patch fixes a regression intoduced by python-2.7.17+-distutils-C++.patch
and updates the distutils testsuite to pass with 2.7.16:2.7.18:
```

##### Detailed analysis
I think this shows the effect of the addition to this patch by this PR in detail:

In order to pass the unit-tests, the `distutils` test suite would have to change as in this diff, when the bug is not fixed:
```diff
         self.assertEqual(comp.exes['compiler'],
-                         'env_cc --sc-cflags --env-cflags --env-cppflags')
+                         'env_cc  --env-cflags --env-cppflags')
         self.assertEqual(comp.exes['compiler_so'],
-                         ('env_cc --sc-cflags '
+                         ('env_cc  '
                           '--env-cflags ''--env-cppflags --sc-ccshared'))
         self.assertEqual(comp.exes['compiler'],
-                         'sc_cc --sc-cflags')
+                         'sc_cc ')
         self.assertEqual(comp.exes['compiler_so'],
-                         'sc_cc --sc-cflags --sc-ccshared')
+                         'sc_cc  --sc-ccshared')
         self.assertEqual(comp.exes['compiler_cxx'],
-                         'sc_cxx')
+                         'sc_cxx ')
```
This means, the currently used `python-2.7.17+-distutils-C++.patch` does not pass the CFLAGS from env-variable (and it seems also not other cflags) into the build of python modules and the new addition by this PR restores that back to what the test suite expects. This could have an effect on python modules when built against python 2.7.16:2.7.18.

The latter change is the only change visible to other packages. It restores the normally assumed handing of `CFLAGS` as it was before `python-2.7.17+-distutils-C++.patch` (which was only intented to add better C++ and `CXXFLAGS` support).

My motivation for this PR was to fix `spack install --test=all` when installing other packages which have python as dependency.
Besides this, there was nothing else prompting this PR.

It's been 14 days now and all the tests in the pipeline passed (it took a while because of the branch name containing `++`)

Do you approve this PR as well?

adamjstewart reviewed (approved):
Based on the fact that the Python 3 patch only modifies the tests (which won't be installed) and Python 2 shouldn't be used anyway, I think it's safe to merge this. If @skosukhin or @xjrc don't get back to you within a couple days let me know and I'll merge this.

bernhardkaindl commented:
> Based on the fact that the Python 3 patch only modifies the tests (which won't be installed) and Python 2 shouldn't be used anyway, I think it's safe to merge this. If @skosukhin or @xjrc don't get back to you within a couple days let me know and I'll merge this.

@adamjstewart Yes, correct! Sergey told me in PR #26149 that he is busy until the 18th October.

@xjrc, You already responded positively:

Is it for you to merge this, given Sergey seems to be too busy for his review?

adamjstewart commented:
Done. Sorry it took so long to get this merged!

bernhardkaindl commented:
> Done. Sorry it took so long to get this merged!

Thanks Adam! No problem. I actually like the level of review done by all of you.

Initially I didn't know what to expect and how it's done, but it's good now :)

To ask Joseph(xjrc) to review the changes and then have a good feedback from him was reassuring that it's OK to proceed.

I didn't anticipate that the the branch name conaining "C++" could become an issue, but that has also been solved and it's all done now, thanks for the review and the merge!
